zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/util/ZkPathRecursiveWatcherTrie.java

```diff
@@ -0,0 +1,271 @@
+package org.apache.helix.zookeeper.zkclient.util;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.helix.zookeeper.zkclient.RecursivePersistListener;
+
+
+/**
+ * ZkPathRecursiveWatcherTrie will be used for a registry for ZK persist recursive watcher.
+ * When persist recursive watcher is registered on path /X, ZK will send out data change for any
+ * data/child changing under the tree structure of /X. The event only include the path of changed
+ * ZNode.
+ * In ZkClient, when ever we get a dataChange event for path /X/Y/Z, we need to track back the path
+ * and notify all registered recursive persist listener and notify them about the node change.
+ * ref: https://zookeeper.apache.org/doc/r3.7.1/zookeeperProgrammers.html#sc_WatchPersistentRecursive
+ */
+public class ZkPathRecursiveWatcherTrie {
+
+  /** Root node of PathTrie */
+  private final TrieNode rootNode;
+
+  private final ReadWriteLock lock = new ReentrantReadWriteLock(true);
+
+  private final Lock readLock = lock.readLock();
+
+  private final Lock writeLock = lock.writeLock();
+
+  static class TrieNode {
+
+    final String value;   // Segmented ZNode path at current level
+    final Map<String, TrieNode> children; // A map of segmented ZNode path of next level to TrieNode
+    TrieNode parent;  // A reference to upper level TrieNode
+    Set<RecursivePersistListener> recursiveListeners;
+    // A list of recursive persist watcher on the current path
+
+    /**
+     * Create a trie node with parent as parameter.
+     *
+     * @param parent the parent of this node
+     * @param value the value stored in this node
+     */
+    private TrieNode(TrieNode parent, String value) {
+      this.value = value;
+      this.parent = parent;
+      this.children =
+          new HashMap<>(4);  // We keep the same number of init children side as Zk sevrer
+      this.recursiveListeners = new HashSet<>(4);
+    }
+
+    /**
+     * Get the parent of this node.
+     *
+     * @return the parent node
+     */
+    TrieNode getParent() {
+      return this.parent;
+    }
+
+    /**
+     * set the parent of this node.
+     *
+     * @param parent the parent to set to
+     */
+    void setParent(TrieNode parent) {
+      this.parent = parent;
+    }
+
```

Comment:
Why you need these two? Usually trie scaning is top down. If the data is missing, we either created it or just skip.

Reply:
TFTR. When deleting, we go from leaf to root. So a reference to parent is needed.

Reply:
There are ways of deleting it without double linked pointer: 1) use recursive call (that seemed match your function name) or 2) use a stack to record the paths.

Combine the replies together. For parallel process, if you have multiple children, you can let each one of them do sub trie deletion in parallel. It is possible for recursive call, because delete operation can be idempotent. 

It could be just listChild.parallelStream().(x -> deleteChild(node)). Instead of passing name, you pass the TrieNode reference.

Reply:
You still keep these two?